crates/wysiwyg/src/composer_model/quotes.rs

```diff
@@ -0,0 +1,319 @@
+
+        self.create_update_replace_all()
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::tests::testutils_composer_model::{cm, tx};
+
+    #[test]
+    fn apply_quote_to_empty_dom() {
+        let mut model = cm("|");
+        model.quote();
+        assert_eq!(tx(&model), "<blockquote>~|</blockquote>")
+    }
+
+    #[test]
+    fn apply_quote_to_simple_text() {
+        let mut model = cm("Some text|");
+        model.quote();
+        assert_eq!(tx(&model), "<blockquote>~Some text|</blockquote>")
+    }
+
+    #[test]
+    fn apply_quote_to_formatted_text() {
+        let mut model = cm("<i>Some text|</i>");
+        model.quote();
+        assert_eq!(tx(&model), "<blockquote>~<i>Some text|</i></blockquote>")
+    }
+
+    #[test]
+    fn apply_quote_to_nested_formatted_text() {
+        let mut model = cm("<i><b>Some text|</b></i>");
+        model.quote();
+        assert_eq!(
+            tx(&model),
+            "<blockquote>~<i><b>Some text|</b></i></blockquote>"
+        )
+    }
+
+    #[test]
+    fn apply_quote_to_text_and_formatted_text() {
+        let mut model = cm("Plain text and <i><b>Some formatted text|</b></i>");
+        model.quote();
+        assert_eq!(
+            tx(&model),
+            "<blockquote>~Plain text and <i><b>Some formatted text|</b></i></blockquote>"
+        )
+    }
+
+    #[test]
+    fn apply_quote_to_single_list_item() {
+        let mut model = cm("<ul><li>~List item|</li></ul>");
+        model.quote();
+        assert_eq!(
+            tx(&model),
+            "<ul><li>~<blockquote>~List item|</blockquote></li></ul>"
+        )
+    }
+
+    #[test]
+    fn apply_quote_to_list() {
+        let mut model =
+            cm("<ul><li>~First {item</li><li>~Second}| item</li></ul>");
+        model.quote();
+        assert_eq!(
+            tx(&model),
+            "<blockquote>~<ul><li>~First {item</li><li>~Second}| item</li></ul></blockquote>"
+        )
+    }
+
+    #[test]
+    fn apply_quote_to_list_and_formatted_text() {
+        let mut model =
+            cm("Plain text <b>bold {text</b><ul><li>~First item</li><li>~Second}| item</li></ul>");
+        model.quote();
+        assert_eq!(
+            tx(&model),
+            "<blockquote>~Plain text <b>bold {text</b><ul><li>~First item</li><li>~Second}| item</li></ul></blockquote>"
+        )
+    }
+
+    #[test]
+    fn apply_quote_to_simple_text_with_line_breaks() {
+        let mut model = cm("Some |text<br />Next line");
+        model.quote();
+        assert_eq!(
+            tx(&model),
+            "<blockquote>~Some |text</blockquote><br />Next line"
+        )
+    }
+
+    #[test]
+    fn apply_quote_to_several_nodes_with_line_breaks() {
+        let mut model = cm("Some {text<br /><b>Next}| line</b>");
+        model.quote();
+        assert_eq!(
+            tx(&model),
+            "<blockquote>~Some {text<br /><b>Next}| line</b></blockquote>"
+        )
+    }
```

Comment:
It looks like the block quote wraps the whole editor content. But if you've selected a part of the content, wouldn't you expect only the selected part to be quoted? I.e.
```suggestion
    fn apply_quote_to_several_nodes_with_line_breaks() {
        let mut model = cm("Some {text<br /><b>Next}| line</b>");
        model.quote();
        assert_eq!(
            tx(&model),
            "Some <blockquote>~{text<br /><b>Next</b>}|</blockquote><b> line</b>"
        )
    }
```

Similarly, on unformatted text wouldn't we expect the following behaviour:
```
    #[test]
    fn apply_quote_to_simple_text_start() {
        let mut model = cm("{Some}| text");
        model.quote();
        assert_eq!(tx(&model), "<blockquote>~Some|</blockquote> text")
    }
```

Reply:
Actually, ignore this. I just tried in a few different editors and now I understand what we're going for.